content-src/asrouter/components/FXASignupForm/FXASignupForm.jsx

```diff
@@ -0,0 +1,187 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { actionCreators as ac } from "common/Actions.jsm";
+import { addUtmParams } from "../../templates/FirstRun/addUtmParams";
+import React from "react";
+
+export class FXASignupForm extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.onSubmit = this.onSubmit.bind(this);
+    this.onInputChange = this.onInputChange.bind(this);
+    this.onInputInvalid = this.onInputInvalid.bind(this);
+    this.handleSignIn = this.handleSignIn.bind(this);
+
+    this.state = {
+      emailInput: "",
+      isSignIn: false,
+    };
+  }
+
+  get email() {
+    return this.props.document
+      .getElementById("fxaSignupForm")
+      .querySelector("input[name=email]");
+  }
+
+  get signInURL() {
+    const urlObj = new URL(this.props.fxaEndpoint);
+    urlObj.pathname = "signin";
+    return urlObj.toString();
+  }
+
+  onSubmit(event) {
+    // Dynamically require the email on submission so screen readers don't read
+    // out it's always required because there's also ways to skip the modal
+    if (this.state.isSignIn) {
+      return;
+    }
+    const { email } = event.target.elements;
+    if (!email.value.length) {
+      email.required = true;
+      email.checkValidity();
+      event.preventDefault();
+      return;
+    }
+
+    // Report to telemetry additional information about the form submission.
+    const value = { has_flow_params: !!this.props.flowParams.flowId.length };
+    this.props.dispatch(ac.UserEvent({ event: "SUBMIT_EMAIL", value }));
+
+    global.addEventListener("visibilitychange", this.props.onClose);
+  }
+
+  handleSignIn(event) {
+    if (event.target.value === "signin") {
+      this.setState({ isSignIn: true });
```

Comment:
Is this setState to trigger a re-render and make the form action point to the appropriate url? If so, you should be able to detect `email.disabled` from `onSubmit` to handle the signin flow. So this `handleSignIn` method should only need to do one thing and the dispatch/visibilitychange behavior can share the existing code in onSubmit, e.g.,:
```js
handleSignIn() {
  this.email.disabled = true;
}
onSubmit(event) {
  let userEvent = "SUBMIT_EMAIL";
  if (this.email.disabled) {
    event.target.setAttribute("action", this.signInURL);
    userEvent = "SUBMIT_SIGNIN";
  } else if (!email.value.length) { … }
  dispatch(…)
```
I believe setting the `<button type="submit">` for signin should trigger the onSubmit if not already.

Reply:
done, no need for explicit type="submit" looks like button click inside a form triggers submit action